controllers/workplacement_controller.go

```diff
@@ -129,42 +133,56 @@ func (r *WorkPlacementReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 }
 
 func (r *WorkPlacementReconciler) deleteWorkPlacement(ctx context.Context, writer writers.StateStoreWriter, workPlacement *v1alpha1.WorkPlacement, filePathMode string, logger logr.Logger) (ctrl.Result, error) {
-	if !controllerutil.ContainsFinalizer(workPlacement, repoCleanupWorkPlacementFinalizer) {
+	pendingRepoCleanup := controllerutil.ContainsFinalizer(workPlacement, repoCleanupWorkPlacementFinalizer)
+	pendingKratixFileCleanup := controllerutil.ContainsFinalizer(workPlacement, kratixFileCleanupWorkPlacementFinalizer)
+
+	if !pendingRepoCleanup && !pendingKratixFileCleanup {
 		return ctrl.Result{}, nil
 	}
-	logger.Info("cleaning up work on repository", "workplacement", workPlacement.Name)
 
 	var err error
-
-	var dir = getDir(*workPlacement) + "/"
 	var workloadsToDelete []string
+	var finalizerToRemove string
+	kratixFilePath := fmt.Sprintf(".kratix/%s-%s.yaml", workPlacement.Namespace, workPlacement.Name)
 
-	if filePathMode == v1alpha1.FilepathModeNone {
-		var kratixFile []byte
-		kratixFilePath := fmt.Sprintf(".kratix/%s-%s.yaml", workPlacement.Namespace, workPlacement.Name)
-		if kratixFile, err = writer.ReadFile(kratixFilePath); err != nil {
-			logger.Error(err, "failed to read .kratix state file")
-			return defaultRequeue, err
-		}
-		stateFile := StateFile{}
-		if err = yaml.Unmarshal(kratixFile, &stateFile); err != nil {
-			logger.Error(err, "failed to unmarshal .kratix state file")
-			return defaultRequeue, err
+	var dir string
+	switch filePathMode {
+	case v1alpha1.FilepathModeNestedByMetadata:
+		dir = getDir(*workPlacement) + "/"
+	}
+
+	if pendingRepoCleanup {
+		finalizerToRemove = repoCleanupWorkPlacementFinalizer
+		logger.Info("cleaning up work on repository", "workplacement", workPlacement.Name)
+		if filePathMode == v1alpha1.FilepathModeNone {
+			var kratixFile []byte
+			if kratixFile, err = writer.ReadFile(kratixFilePath); err != nil {
+				logger.Error(err, "failed to read .kratix state file", "file path", kratixFilePath)
+				return ctrl.Result{}, err
+			}
+			stateFile := StateFile{}
+			if err = yaml.Unmarshal(kratixFile, &stateFile); err != nil {
+				logger.Error(err, "failed to unmarshal .kratix state file")
+				return defaultRequeue, err
+			}
+			workloadsToDelete = stateFile.Files
 		}
-		dir = ""
-		workloadsToDelete = append(stateFile.Files, kratixFilePath)
 	}
-	_, err = writer.UpdateFiles(dir, workPlacement.Name, nil, workloadsToDelete)
 
-	if err != nil {
+	if !pendingRepoCleanup && pendingKratixFileCleanup {
+		logger.Info("cleaning up .kratix state file", "workplacement", workPlacement.Name)
+		workloadsToDelete = []string{kratixFilePath}
+		finalizerToRemove = kratixFileCleanupWorkPlacementFinalizer
+	}
+
+	if _, err := writer.UpdateFiles(dir, workPlacement.Name, nil, workloadsToDelete); err != nil {
```

Comment:
took me way to long to grok how this worked 🙈 i think codes a lot more complex just to reduce a bit of duplication. I'd rather it was just duplicated and looked ike:

```
if repoFinalierCleanuper
   ... logic
  remove finalizer
  requeue

if pendingRepoCleanup
   ... logic
   remove finalizer
   requeue
```

Reply:
i had something like this before but didn't like the duplication, hence the current state;

i could extract it to a function though... let me give it a go